fix: added a prerequisite check to ensure certain conditions are fulfilled before running docker compose build and watch

sidtohan opened the pull request:
## Linked Issue(s) 

fixes #338

## Acceptance Criteria fulfillment

- [x] Check if the required ports are available (list the specific ports needed)
- [x] Check if the Docker daemon is running
- [x] Check if the docker-compose.yml file exists in the project directory
- [x] Check if the Dockerfile.dev exists in the project directory
- [x] Add suggestions or solutions for each error
- [x] Update UI to display errors properly
- [x] Check if the docker version is the correct version which supports `docker compose watch`
- [ ] Add unit tests

## Proposed changes (including videos or screenshots)
- Created a custom hook that checks if docker daemon is running, if the required ports are free, if docker-compose.yml file exists and dockerfile.dev exists.
- Created a basic UI that lists down all the checks when they are running, and if any of them fails, it terminates the process
![image](https://github.com/middlewarehq/middleware/assets/82583557/866b9a45-1d14-4c56-8561-f0af23913196)

<!-- CHANGELOG -->

- Added a custom app state called `PREREQ_CHECK`,  which denotes that the app is currently checking the prerequisites. Also, the initial state of the app is now `PREREQ_CHECK` instead of `INIT`.
- Added a new enum in `cli/source/constants.ts` called `PreCheckStates` consisting of three values: `RUNNING, SUCCESS` and `FAILED`. 

<!-- END CHANGELOG -->


## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->
There are currently a lot of major changes regarding the dependency arrays for the `useEffect` and the `useCallback` hooks. I have currently implemented them as per my opinion, but I would appreciate some guidance on how we can make it look more cleaner and better.

There is currently a lot of work to be done on the UI as well, but I would first like to get some feedback on the type of approach I've taken. Thanks.

jayantbh commented:
Great suggestion, @sidtohan!
Thanks for your PR. We'll look into this.

jayantbh commented:
Btw, did you run `npm install` by any chance?
Because `cli/package-lock.json` should not be a new file. You'll see there should be a `yarn.lock` indicating that it should use `yarn` as the package manager.

As a result, your PR blew to 7.6k lines. :)

jayantbh reviewed (changes requested):
This is an incomplete review. A more thorough review would be done by someone else from the team.

jayantbh commented:
Hey @sidtohan, are you taking care of the package-lock.json thing?

sidtohan commented on cli/source/app.tsx:
```diff
@@ -149,27 +162,57 @@ const CliUi = () => {
-      ['compose', 'build'],
-      runCommandOpts
-    );
+    if (
+      preCheck.daemon !== PreCheckStates.SUCCESS ||
+      preCheck.ports !== PreCheckStates.SUCCESS ||
+      preCheck.composeFile !== PreCheckStates.SUCCESS ||
+      preCheck.dockerFile !== PreCheckStates.SUCCESS
+    ) {
+      if (
+        preCheck.daemon === PreCheckStates.FAILED ||
+        preCheck.ports === PreCheckStates.FAILED ||
+        preCheck.composeFile === PreCheckStates.FAILED ||
+        preCheck.dockerFile === PreCheckStates.FAILED
+      )
+        handleExit();
+      return;
+    }
```
Actually what I've tried to do here is that if any of the checks is in failed state, then I need to quit the program.
But if any of checks is currently in progress ( essentially the `PreCheckStates.RUNNING` state), then I want to return from here but not exit the program. But you are right, it is quite poorely written. I'll try to simplify it further.

sidtohan commented:
> Hey @sidtohan, are you taking care of the package-lock.json thing?

Yes, I'm working on it, sorry for the extremely slow response.

sidtohan commented:
> <img alt="image" width="533" src="https://private-user-images.githubusercontent.com/7949047/346076540-8dd56d98-a860-41bb-a211-f232a6229af7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAyODEyNjQsIm5iZiI6MTcyMDI4MDk2NCwicGF0aCI6Ii83OTQ5MDQ3LzM0NjA3NjU0MC04ZGQ1NmQ5OC1hODYwLTQxYmItYTIxMS1mMjMyYTYyMjlhZjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDZUMTU0OTI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzA0OGNiODlhZGY2YzU4YjI2M2Y4MTVmMWFjOTQ3YTI4MTc1ZjA1Njg5YTM1ZWQzYThiY2RlN2EzMjRiM2NlMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ._dSeYA3ltoWilZJQjleSSBy7Nobrs78BU_Z-ysOpQ5U">
> 
> Your yarn.lock changes would need to be reset as well. It looks like you changed the registry URLs.
> 
> @sidtohan

Sure I'll reset them.

sidtohan commented:
So I've added the requested changes, regarding the check for the states of the precheck objects, I would greatly appreciate any inputs on how to improve further. 

I've removed the package-lock file and added the installation of the module "find-free-ports" which I'm using in the yarn.lock file. Please let me know if there is any issue regarding that.

Kindly review this whenever convenient. I'll try to be more responsive next week.

shivam-bit commented on cli/source/app.tsx:
```diff
@@ -264,12 +299,37 @@ const CliUi = () => {
+    property
+  }: {
+    value: PreCheckStates;
+    property: string;
```
@sidtohan can you declare type for `property`

sidtohan replied:
Sure, I'll add an enum for this.

samad-yar-khan commented:
@sidtohan Great PR! Although ports are blocked and daemon is also closed. Its only showing that ports are not accessible and nothing about the daemon. only showing one issue at a time.
![image](https://github.com/middlewarehq/middleware/assets/70485812/9a5741b1-40cc-4425-821f-6c5ab1809056)

![image](https://github.com/middlewarehq/middleware/assets/70485812/4efe7973-b95f-4531-91e0-fb90a4e86549)

samad-yar-khan commented on cli/source/hooks/usePreCheck.ts:
```diff
@@ -0,0 +1,80 @@
+    } else {
+      const db_check = await isFreePort(db);
+      const redis_check = await isFreePort(redis);
+      const fe_check = await isFreePort(frontend);
+      const ss_check = await isFreePort(sync_server);
+      const as_check = await isFreePort(analytics_server);
```
variable names are ambiguous

shivam-bit commented on cli/source/hooks/usePreCheck.ts:
```diff
@@ -0,0 +1,80 @@
+        setDaemon(PreCheckStates.FAILED);
+      });
+
+    // For ports
+    if (
+      isNaN(db) ||
+      isNaN(redis) ||
+      isNaN(frontend) ||
+      isNaN(sync_server) ||
+      isNaN(analytics_server)
+    ) {
+      setPorts(PreCheckStates.FAILED);
+    } else {
+      const db_check = await isFreePort(db);
+      const redis_check = await isFreePort(redis);
+      const fe_check = await isFreePort(frontend);
+      const ss_check = await isFreePort(sync_server);
+      const as_check = await isFreePort(analytics_server);
+      if (!db_check || !redis_check || !fe_check || !ss_check || !as_check) {
+        setPorts(PreCheckStates.FAILED);
+      } else {
+        setPorts(PreCheckStates.SUCCESS);
+      }
+    }
+
+    // For files
+    fs.promises
+      .access('../docker-compose.yml')
+      .then(() => {
+        setComposeFile(PreCheckStates.SUCCESS);
+      })
+      .catch((err) => {
+        setComposeFile(PreCheckStates.FAILED);
+      });
+
+    fs.promises
+      .access('../Dockerfile.dev')
+      .then(() => setDockerFile(PreCheckStates.SUCCESS))
+      .catch(() => setDockerFile(PreCheckStates.FAILED));
+  };
```
I think you should divide this `callChecks` into 3 smaller callbacks
also prefer to enclose those functions into `useCallback`.

adnanhashmi09 commented on cli/source/hooks/usePreCheck.ts:
```diff
@@ -0,0 +1,80 @@
+      });
+
+    // For ports
+    if (
+      isNaN(db) ||
+      isNaN(redis) ||
+      isNaN(frontend) ||
+      isNaN(sync_server) ||
+      isNaN(analytics_server)
+    ) {
+      setPorts(PreCheckStates.FAILED);
+    } else {
+      const db_check = await isFreePort(db);
+      const redis_check = await isFreePort(redis);
+      const fe_check = await isFreePort(frontend);
+      const ss_check = await isFreePort(sync_server);
+      const as_check = await isFreePort(analytics_server);
+      if (!db_check || !redis_check || !fe_check || !ss_check || !as_check) {
+        setPorts(PreCheckStates.FAILED);
+      } else {
+        setPorts(PreCheckStates.SUCCESS);
+      }
+    }
```
You can use `Promise.all()` and `Promise.allSettled()` to avoid having to store values of promises in variables and conditionally checking them. It might make code more readable, imo. Just a suggestion!

sidtohan replied:
So here what would be the better option between `Promise.all()` and `Promise.allSettled()`? If we want better display of failed checks, then it might be better to also display the exact port where it fails (unlike what is being done in the current UI that I implemented). Would the `Promise.allSettled()` be better in that case as it would run all the checks even if one of them fails?

jayantbh replied:
Definitely allSettled.
I don't know about readability, but it'll def be faster.

My suggestion:

```js
const portPromises = [db, redis, frontend, sync_server, analytics_server].map(isFreePort);
const checks = await Promise.allSettled(portPromises);
if (checks.includes(false)) failed;
else passed;
```

sidtohan commented on cli/source/app.tsx:
```diff
@@ -264,12 +299,37 @@ const CliUi = () => {
+          <Text color="green">✓</Text>
+        ) : (
+          <Text color="red">x</Text>
+        )}{' '}
```
I added this for some space after the tick and cross. If there is some other better way for this please let me know.

shivam-bit commented:
LGTM
@adnanhashmi09  please check if your requested changed were address or not

adnanhashmi09 commented:
@sidtohan could you also add this check
- [ ] Check if the docker version is the correct version which supports docker compose watch

It was added in the PR description as well

samad-yar-khan commented:
Hey @sidtohan, the port checking logic is buggy. If you block one of your ports, you will able to see that the CLI dies not show the error message for ports. 

![image](https://github.com/user-attachments/assets/1b00375f-a420-48b0-8db7-1248593d09cf)

The error is likely due to the changes made in this [commit](https://github.com/middlewarehq/middleware/pull/451/commits/dcd5c6e626ae8e4a23b9ea0efecb2b05a07b8fe7). Basically, the code was updated to look leaner but the end logic only checks for the promises being settled in `callPortsCheck` and not the values.

![image](https://github.com/user-attachments/assets/48faa1d2-212e-4805-89d0-5cdca74ab69c)

The below code change should fix the issue.

`      if (checks.some((item) => item.status === 'rejected' || item.value === false)) {`

samad-yar-khan commented on cli/source/hooks/usePreCheck.ts:
```diff
@@ -0,0 +1,84 @@
+    } else {
+      const portPromises = ports_array.map(isFreePort);
+      const checks = await Promise.allSettled(portPromises);
+      if (checks.some((item) => item.status === 'rejected')) {
```
This should fix the bug in the above comment @sidtohan

sidtohan replied:
hey @samad-yar-khan , I've made the requested changes. Could you please review it once? Thanks!

samad-yar-khan reviewed (changes requested):
@sidtohan  Everything works as expected now! but not able to merge due to conflicts. Can you please rebase the branch ?

samad-yar-khan reviewed (approved):
Amazing work @sidtohan ! Thanks for this contribution 🎉  🚀

sidtohan commented:
Happy to contribute! 🥳

jayantbh commented:
Just wanted to drop a comment to appreciate that this was a really solid contribution. 👏🏽 👏🏽 
Definitely avoids certain confusions when running the dev setup. Good stuff, @sidtohan 

In case you're not on our Slack, feel free to join!
https://join.slack.com/t/middle-out-group/shared_invite/zt-2hbuoog3a-k~ojcay1yDJ~a2kVPeEMJg

We'd love to build more things with you.
